### PR TITLE
feat: Automatically prefer localized pathnames that are more specific

### PR DESCRIPTION
**Pull request by fkapsahili**
This PR fixes the current behavior of the internal route handling of the `next-intl` middleware that does not prioritize static route paths over dynamic routes. This fix makes sure that static routes (e.g. `/products/new`) are preferred over dynamic routes such as `/products/[slug]`.

To make sure the current behavior does not lead to any unexpected issues, I thought it would be good if `next-intl` would prioritize the specified `pathnames` in the following order:
1. static pathnames, e.g. `/products/new`
2. dynamic routes (non-catch-all), e.g. `/products/[slug]`
3. catch-all routes, e.g. `/products/[...slug]` or optional catch-all routes, e.g. `/products/[[...slug]]`

### Prerequisites

- [x] Use a meaningful title for the pull request
- [x] Test the changes you've made by adding or editing tests

### Content
- Added `getSortedPathnames` to handle the sorting of the specified pathnames in the `middleware`
- Updated the `docs` to make sure the changes are reflected in the documentation

#### Tests
- Added tests to the `middleware` that verify that static routes are prioritized over dynamic routes and catch-all routes
- Added tests to the middleware `utils` to make sure that `getSortedPathnames` handles the route prioritization as intended
- Added playwright tests to the `example-app-router-playground` to make sure the route prioritization works as intended 

Closes https://github.com/amannn/next-intl/issues/913

### CODE REVIEW

**Review by amannn — COMMENTED** [review]
Many thanks for looking into this and another great PR @fkapsahili! 👏👏

On a first glance, this looks like just the right implementation. I have a somewhat busy week ahead, but I'll likely be able to take a thorough look on Friday—hope that works for you!

I'll be in touch! 🙌

**Comment by fkapsahili** [comment]
> Many thanks for looking into this and another great PR @fkapsahili! 👏👏
> 
> On a first glance, this looks like just the right implementation. I have a somewhat busy week ahead, but I'll likely be able to take a thorough look on Friday—hope that works for you!
> 
> I'll be in touch! 🙌

Thanks for the feedback! Yes of course, that works well for me 🙂. The tests are all working now, so I'm setting the PR to ready for review 😄.

File: packages/next-intl/src/middleware/utils.tsx [hunk 1]
```diff
@@ -9,6 +9,71 @@ export function getFirstPathnameSegment(pathname: string) {
...
+  a: PathnamePair<Locales>,
+  b: PathnamePair<Locales>
+): number {
+  const pathA = getExternalPath(a[1]);
+  const pathB = getExternalPath(b[1]);
+
+  const dynamicA = isDynamicRoute(pathA);
+  const dynamicB = isDynamicRoute(pathB);
+  const catchAllA = isCatchAllRoute(pathA);
+  const catchAllB = isCatchAllRoute(pathB);
```
Comment (amannn) [inline packages/next-intl/src/middleware/utils.tsx:50]: Diving a bit deeper into this, I'm not sure these flags are sufficient to be honest.

E.g. consider this case:
```
/articles/[category]/[articleSlug]
/articles/[category]/just-in
```

In this case, the routes would be considered equal in regard to priority, and probably the natural ordering would be preserved. However, the second route is more specific and should match.

The number of dynamic segments could be another indicator, but I think this won't solve the problem cleanly as e.g. `/foo/[d1]` and `/foo/bar/[d1]` have the same number of segments.

It seems like to be able to tell which route is really more specific, we have to consider routes in a hierarchical way, e.g. find the most specific route in `/articles`, then try to find the most specific route in `./[category]`, etc.

Do you by chance know if Next.js handles these cases correctly? Maybe we can take some inspiration from them.

What do you think?

File: packages/next-intl/test/middleware/utils.test.tsx [hunk 2]
```diff
@@ -162,3 +163,85 @@ describe('getInternalTemplate', () => {
     ]);
   });
 });
+
+describe('getSortedPathnames', () => {
```
Comment (amannn) [inline packages/next-intl/test/middleware/utils.test.tsx:167]: Good idea to add some unit tests for this function!

File: packages/next-intl/src/middleware/utils.tsx (hunk 1 above)
Comment (amannn) [reply packages/next-intl/src/middleware/utils.tsx:50]: Or maybe the number of dynamic segments could in fact suffice? 🤔 

In the example `/foo/[d1]` and `/foo/bar/[d1]`, the routes have the same priority, but only one of them matches eventually—so maybe this isn't an issue.

Do you think you could a few more tests for edge cases so we can gain some clarity about those?

Comment (fkapsahili) [reply packages/next-intl/src/middleware/utils.tsx]: Ah yes, thanks for the hint! You were absolutely right, my solution didn't take into account the case where we have nested dynamic routes, e.g: `/articles/[category]/[slug]` and a static route in the same place 😅.

I'm not very familiar with the Next.js repo, but at first glance I believe the code you referenced in the corresponding issue is actually where Next does the route sorting:
https://github.com/vercel/next.js/blob/415cd74b9a220b6f50da64da68c13043e9b02995/packages/next/src/shared/lib/router/utils/sorted-routes.ts

When looking at Next's code I realized that it is necessary that the individual route segments are compared with each other and have now implemented this additionally so that the failed tests that I added at the beginning of this update now pass.

If there are any other edge cases you can think of, I'm happy to add more tests! 🙂

File: packages/next-intl/src/middleware/utils.tsx [hunk 3]
```diff
@@ -19,10 +87,13 @@ export function getInternalTemplate<
   pathname: string,
   locale: Locales[number]
 ): [Locales[number] | undefined, keyof Pathnames | undefined] {
+  // Sort pathnames by specificity
+  const sortedPathnames = getSortedPathnames(pathnames);
```
Comment (amannn) [inline packages/next-intl/src/middleware/utils.tsx:91]: Would suggest to add a line break before the next comment just to bikeshed a bit about code aesthetics 😄 

```suggestion
  const sortedPathnames = getSortedPathnames(pathnames);

```

File: packages/next-intl/test/middleware/middleware.test.tsx [hunk 4]
```diff
@@ -642,6 +672,54 @@ describe('prefix-based routing', () => {
         );
       });
 
+      it('priotizes static routes over dynamic and catch-all routes for the non-default locale', () => {
```
Comment (amannn) [inline packages/next-intl/test/middleware/middleware.test.tsx:675]:
```suggestion
      it('prioritizes static routes over dynamic and catch-all routes for the non-default locale', () => {
```

File: packages/next-intl/test/middleware/middleware.test.tsx [hunk 5]
```diff
@@ -642,6 +672,54 @@ describe('prefix-based routing', () => {
...
+        expect(MockedNextResponse.rewrite).toHaveBeenCalledTimes(2);
+        expect(MockedNextResponse.rewrite.mock.calls[0][0].toString()).toBe(
+          'http://localhost:3000/en/products/add'
+        );
+        expect(MockedNextResponse.rewrite.mock.calls[1][0].toString()).toBe(
+          'http://localhost:3000/en/categories/new'
+        );
+      });
+
+      it('priotizes static routes over dynamic and catch-all routes for the non-default locale', () => {
```
Comment (amannn) [inline packages/next-intl/test/middleware/middleware.test.tsx:689]:
```suggestion
      it('prioritizes static routes over dynamic and catch-all routes for the non-default locale', () => {
```

File: packages/next-intl/test/middleware/middleware.test.tsx [hunk 6]
```diff
@@ -642,6 +672,54 @@ describe('prefix-based routing', () => {
...
+        expect(MockedNextResponse.rewrite).toHaveBeenCalledTimes(2);
+        expect(MockedNextResponse.rewrite.mock.calls[0][0].toString()).toBe(
+          'http://localhost:3000/de/products/add'
+        );
+        expect(MockedNextResponse.rewrite.mock.calls[1][0].toString()).toBe(
+          'http://localhost:3000/de/categories/new'
+        );
+      });
+
+      it('priotizes more specific, static routes over dynamic routes for the non-default locale', () => {
```
Comment (amannn) [inline packages/next-intl/test/middleware/middleware.test.tsx:705]:
```suggestion
      it('prioritizes more specific, static routes over dynamic routes for the non-default locale', () => {
```

File: packages/next-intl/test/middleware/middleware.test.tsx [hunk 7]
```diff
@@ -1100,10 +1198,76 @@ describe('prefix-based routing', () => {
...
+        expect(MockedNextResponse.rewrite).toHaveBeenCalledTimes(2);
+        expect(MockedNextResponse.rewrite.mock.calls[0][0].toString()).toBe(
+          'http://localhost:3000/en/products/add'
+        );
+        expect(MockedNextResponse.rewrite.mock.calls[1][0].toString()).toBe(
+          'http://localhost:3000/en/categories/new'
+        );
+      });
+
+      it('priotizes more specific, static routes over dynamic routes for the default locale', () => {
```
Comment (amannn) [inline packages/next-intl/test/middleware/middleware.test.tsx:1219]:
```suggestion
      it('prioritizes more specific, static routes over dynamic routes for the default locale', () => {
```

**Review by amannn — COMMENTED** [review]
Great refactor, this looks really solid to and I think this is the right approach 👏.

I've left some inline comments about minor nits, but I also think there might be a chance to simplify this function slightly.

Let me know what you think! 🙌

File: packages/next-intl/src/middleware/utils.tsx [hunk 8]
```diff
@@ -9,6 +9,74 @@ export function getFirstPathnameSegment(pathname: string) {
...
+  string,
+  string | {[Key in Locales[number]]: string}
+];
+
+export function comparePathnamePairs<Locales extends AllLocales>(
+  a: PathnamePair<Locales>,
+  b: PathnamePair<Locales>
+): number {
+  const pathA = getExternalPath(a[1]).split('/').filter(Boolean);
+  const pathB = getExternalPath(b[1]).split('/').filter(Boolean);
```
[the remark this answers is not in the document]
Comment (fkapsahili) [reply packages/next-intl/src/middleware/utils.tsx:41]: Ah yes, that makes sense! I think the `getExternalPath` comes from a first attempt where I tried to separate everything very strictly. Just taking the internal route name is a good idea! I'll try to clean that up a bit so we have an easier to read test suite and a more minimalistic approach in the utils 👍🏼.

**Comment by fkapsahili** [comment]
@amannn I have now simplified the sorting function of `getSortedPathnames` so that only the internal pathnames are passed as arguments, which makes the tests look a bit easier. Just let me know if there is anything I can improve!

**Review by amannn — APPROVED** [review]
Looks fantastic, a really cool quality-of-life improvement! 👏 

I'll just add one commit with a minor renaming nit I spotted during the final review.
